pkg/sentry/fsimpl/erofs/erofs.go

```diff
@@ -0,0 +1,362 @@
+
+	// Immutable options.
+	mopts string
+
+	// devMinor is the filesystem's minor device number. devMinor is immutable.
+	devMinor uint32
+
+	// root is the root dentry. root is immutable.
+	root *dentry
+
+	// image is the EROFS image. image is immutable.
+	image *erofs.Image
+}
+
+// Name implements vfs.FilesystemType.Name.
+func (FilesystemType) Name() string {
+	return Name
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (FilesystemType) Release(ctx context.Context) {}
+
+// GetFilesystem implements vfs.FilesystemType.GetFilesystem.
+func (fstype FilesystemType) GetFilesystem(ctx context.Context, vfsObj *vfs.VirtualFilesystem, creds *auth.Credentials, source string, opts vfs.GetFilesystemOptions) (*vfs.Filesystem, *vfs.Dentry, error) {
+	mopts := vfs.GenericParseMountOptions(opts.Data)
+
+	fd, err := getFDFromMountOptionsMap(ctx, mopts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	f := os.NewFile(uintptr(fd), "image file")
+	image, err := erofs.OpenImage(f)
+	if err != nil {
+		f.Close()
+		return nil, nil, err
+	}
+
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		image.Close()
+		return nil, nil, err
+	}
+
+	fs := &filesystem{
+		mopts:    opts.Data,
+		image:    image,
+		devMinor: devMinor,
+	}
+	fs.vfsfs.Init(vfsObj, &fstype, fs)
+
+	root := fs.newDentry(image.Root())
+	// Increase the root's reference count to 2. One reference is returned to
+	// the caller, and the other is held by fs.
+	root.IncRef()
+	fs.root = root
+
+	return &fs.vfsfs, &root.vfsd, nil
+}
+
+func getFDFromMountOptionsMap(ctx context.Context, mopts map[string]string) (int, error) {
+	ifdstr, ok := mopts[moptImageFD]
+	if !ok {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: image FD must be specified as '%s=<file descriptor>'", moptImageFD)
+		return -1, linuxerr.EINVAL
+	}
+	delete(mopts, moptImageFD)
+
+	ifd, err := strconv.Atoi(ifdstr)
+	if err != nil {
+		ctx.Warningf("erofs.getFDFromMountOptionsMap: invalid image FD: %s=%s", moptImageFD, ifdstr)
+		return -1, linuxerr.EINVAL
+	}
+
+	return ifd, nil
+}
+
+// dentry implements vfs.DentryImpl.
+//
+// +stateify savable
+type dentry struct {
+	vfsd vfs.Dentry
+
+	dentryRefs
+
+	// fs is the owning filesystem.
+	fs *filesystem
+
+	// parent is this dentry's parent directory. If this dentry is
+	// a file system root, parent is nil.
+	parent *dentry
+
+	// name is this dentry's name in its parent. If this dentry is
+	// a file system root, name is the empty string.
+	name string
+
+	// inode is the inode represented by this dentry in image.
+	inode *erofs.Inode
+
+	// dirMu serializes changes to the dentry tree.
+	dirMu sync.RWMutex `state:"nosave"`
+
+	// childMap contains the mappings of child names to dentries if
+	// this dentry represents a directory.
+	// +checklocks:dirMu
+	childMap map[string]*dentry
+
+	// mf implements memmap.File for mappings of this dentry.
+	mf dentryMemmapFile
+
+	locks vfs.FileLocks
+
+	// Inotify watches for this dentry.
+	watches vfs.Watches
```

Comment:
These should all be properties of the inode right? Since multiple dentries can point to the same Inode (hard links). EROFS allows hard links right? Otherwise inotify and file locks will not work correctly for hard links.

I think you will need to add another Inode struct in fsimpl/erofs:
```
type inode struct {
     diskInode erofs.Inode
     ...
}
```

And all these fields should come inside that. You should move the `dirents` cache from `erofs.Inode` to this struct too. The `erofs.Inode` should be a simple struct that you can pass by value (currently it is not, because of the presence of `sync.Mutex` in it). But after removing the dirents cache from there (as proposed in the above suggestions), it becomes a simple struct that can be returned by value from dirInode.Lookup().

This inode should also be reference counted. The reference model should be as follwoing:
- Dentry takes a ref on its inode.
- Inode takes a ref on its parent, if available.

Reply:
Good point. Thanks for catching that. I missed it. My previous thoughts were to eliminate the inode struct in `pkg/sentry/fsimpl/erofs` as we already have one in `pkg/erofs`. But some fields like above ones that should stay in `inode` cannot be moved to the inode struct in `pkg/erofs`. I will fix it in the next version. I do like the point that the `erofs.Inode` should be a simple struct that can be passed by value. Thanks for the suggestion!